server/src/main/java/org/elasticsearch/index/mapper/DynamicTemplate.java

```diff
@@ -181,21 +181,24 @@ static DynamicTemplate parse(String name, Map<String, Object> conf) throws Mappe
                 throw new IllegalArgumentException("Illegal dynamic template parameter: [" + propName + "]");
             }
         }
-
-        if (match == null && pathMatch == null && matchMappingType == null) {
-            throw new MapperParsingException("template must have match, path_match or match_mapping_type set " + conf.toString());
-        }
         if (mapping == null) {
             throw new MapperParsingException("template [" + name + "] must have either mapping or runtime set");
         }
 
-        XContentFieldType xcontentFieldType = null;
-        if (matchMappingType != null && matchMappingType.equals("*") == false) {
-            xcontentFieldType = XContentFieldType.fromString(matchMappingType);
-            if (runtime && xcontentFieldType.supportsRuntimeField() == false) {
+        XContentFieldType[] xContentFieldTypes = new XContentFieldType[0];
+        if (matchMappingType == null) {
+            if (match != null || pathMatch != null) {
+                xContentFieldTypes = XContentFieldType.values();
+            }
+        } else if (matchMappingType.equals("*")) {
+            xContentFieldTypes = XContentFieldType.values();
```

Comment:
I wonder with the new logic, if it is counter intuitive that we add all types, knowing that runtime mappings are only applied to a subset of types. Maybe we should filter the types here in that case to only the ones that support runtime mappings? Then we shoud be able to remove the last condition from the match method as well as some logic from RootObjectMapper#validateDynamicTemplate? Do you think it makes sense?